pkg/email/mailgun.go

```diff
@@ -0,0 +1,156 @@
+package email
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"html/template"
+	"io/ioutil"
+	"strings"
+	"time"
+
+	"github.com/juju/errors"
+	"github.com/mailgun/mailgun-go/v4"
+	"gopkg.in/yaml.v2"
+)
+
+// MailgunProviderName is mailgun provider name
+const MailgunProviderName = "mailgun"
+
+const defaultHTMLTemplateForUserEmail = `
+<h1>
+    Here is your 6 digit code.
+    <h2>{{.Code}}</h2>
+</h1>
+`
+
+const defaultHTMLTemplateForContactEmail = `
+<h1>
+    Hello. Here is your link.
+    <h2>{{.Link}}</h2>
+</h1>
+`
+
+// MailgunClientConfig config for Mailgun
+type MailgunClientConfig struct {
+	// ProviderID specifies email provider id
+	ProviderID string `json:"provider_id" yaml:"provider_id"`
+	// Domain specifies domain emails are sent from
+	Domain string `json:"domain" yaml:"domain"`
+	// Sender specifies sender email
+	Sender string `json:"sender" yaml:"sender"`
+	// PrivateAPIKey specifies mailgun private key
+	PrivateAPIKey string `json:"private_api_key" yaml:"private_api_key"`
+}
+
+// MailgunClient client for Mailgun
+type MailgunClient struct {
+	cfg *MailgunClientConfig
+}
+
+// NewMailgunClient creates a new mailgun client
+func NewMailgunClient(cfg *MailgunClientConfig) *MailgunClient {
+	return &MailgunClient{
+		cfg: cfg,
+	}
+}
+
+// SendUserEmail implements email sender for users
+func (m *MailgunClient) SendUserEmail(
+	to string,
+	code string,
+) error {
+	mg := mailgun.NewMailgun(m.cfg.Domain, m.cfg.PrivateAPIKey)
+	message := mg.NewMessage(m.cfg.Sender, "Here is your 6 digit code", "", to)
+
+	tmpl, err := template.New("user-email").Parse(defaultHTMLTemplateForUserEmail)
+	if err != nil {
+		return fmt.Errorf("failed to parse HTML template, err=[%v]", err)
+	}
+
+	var buf bytes.Buffer
+
+	if err := tmpl.Execute(&buf, struct {
+		Code string
+	}{
+		Code: code,
+	}); err != nil {
+		return fmt.Errorf("failed to render HTML template, err=[%v]", err)
+	}
+
+	message.SetHtml(buf.String())
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
+	defer cancel()
+
+	// Send the message with a 10 second timeout
+	_, _, err = mg.Send(ctx, message)
+
+	if err != nil {
+		return fmt.Errorf("failed to send email using mailgun provider: domain=%q, sender=%q, to=%q, err=[%v]", m.cfg.Domain, m.cfg.Sender, to, err)
+	}
+
+	return nil
+}
+
+// SendContactEmail implements email sender for org contacts
+func (m *MailgunClient) SendContactEmail(
```

Comment:
this package shall have only generic Send method.
the message should be rendered outside in martini service